### PR TITLE
DB as a service: remove database deployment with API

### PR DESCRIPTION
**Pull request by odeke-em**
Fixes https://github.com/tendermint/tendermint/issues/1162

Databases as a service!

Can now access Databases as a remote service
via gRPC for performance and easy deployment.

The caveat is that each service is stateful in regards to
the DB i.e. each unique service uses only one unique DB
but nonetheless multiple clients can access it.

A full standalone/end-to-end example in a test

```go
package remotedb_test

import (
	"net"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/tendermint/tmlibs/grpcdb"
	"github.com/tendermint/tmlibs/remotedb"
)

func TestRemoteDB(t *testing.T) {
	ln, err := net.Listen("tcp", "0.0.0.0:0")
	require.Nil(t, err, "expecting a port to have been assigned on which we can listen")
	srv := grpcdb.NewServer()
	defer srv.Stop()
	go func() {
		if err := srv.Serve(ln); err != nil {
			t.Fatalf("BindServer: %v", err)
		}
	}()

	client, err := remotedb.NewInsecure(ln.Addr().String())
	require.Nil(t, err, "expecting a successful client creation")
	require.Nil(t, client.InitRemote(&remotedb.Init{Name: "test-remote-db", Type: "leveldb"}))

	k1 := []byte("key-1")
	v1 := client.Get(k1)
	require.Equal(t, 0, len(v1), "expecting no key1 to have been stored")
	vv1 := []byte("value-1")
	client.Set(k1, vv1)
	gv1 := client.Get(k1)
	require.Equal(t, gv1, vv1)

	// Deletion
	client.Delete(k1)
	gv2 := client.Get(k1)
	require.Equal(t, len(gv2), 0, "after deletion, not expecting the key to exist anymore")
	require.NotEqual(t, len(gv1), len(gv2), "after deletion, not expecting the key to exist anymore")
}
```

### CODE REVIEW

File: grpcdb/server.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+func (s *server) Iterator(query *protodb.Entity, dis protodb.DB_IteratorServer) error {
+	it := s.db.Iterator(query.Start, query.End)
+	return s.handleIterator(it, dis.Send)
+}
+
+func (s *server) handleIterator(it db.Iterator, sendFunc func(*protodb.Iterator) error) error {
+	for it.Valid() {
+		start, end := it.Domain()
+		out := &protodb.Iterator{
+			Domain: &protodb.DDomain{Start: start, End: end},
```
Comment (melekes) [inline grpcdb/server.go]: `DDomain` as Database Domain?

File: proto/defs.proto [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+message Entity {
+  int32 id	= 1;
+  bytes key	= 2;
+  bytes value	= 3;
+  bool exists	= 4;
+  bytes start	= 5;
+  bytes end	= 6;
+  string err	= 7;
+  string print	= 8;
+  int64 time_at = 9;
```
Comment (melekes) [inline proto/defs.proto]: does `time_at` mean `created_at` or `received_at` or ?

File: proto/defs.proto [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+message Entity {
+  int32 id	= 1;
+  bytes key	= 2;
+  bytes value	= 3;
+  bool exists	= 4;
+  bytes start	= 5;
+  bytes end	= 6;
+  string err	= 7;
+  string print	= 8;
```
Comment (melekes) [inline proto/defs.proto]: seems to be unused

**Comment by melekes** [comment]
Looks 🆒

File: proto/defs.proto (hunk 2 above)
Comment (cwgoes) [reply proto/defs.proto]: `created_at`, updated.

File: proto/defs.proto (hunk 3 above)
Comment (cwgoes) [reply proto/defs.proto]: Removed.

File: grpcdb/server.go (hunk 1 above)
Comment (cwgoes) [reply grpcdb/server.go]: Changed to just `Domain`.

**Comment by cwgoes** [comment]
(new issue - ref https://github.com/tendermint/tmlibs/issues/203)

File: grpcdb/client.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+type Security uint
+
+const (
+	Insecure Security = iota
+	Secure
+)
+
+// NewClient creates a gRPC client connected to the bound gRPC server at serverAddr.
+// Use kind to set the level of security to either Secure or Insecure.
+func NewClient(serverAddr string, serverCert string) (protodb.DBClient, error) {
```
Comment (melekes) [inline grpcdb/client.go]: `serverAddr, serverCert string`

File: grpcdb/doc.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	defer srv.Stop()
+	go func() {
+		if err := srv.Serve(ln); err != nil {
+			t.Fatalf("BindServer: %v", err)
+		}
+	}()
+
+or
+	addr := ":8998"
+  cert := "server.crt"
```
Comment (melekes) [inline grpcdb/doc.go]: something wrong with formatting

File: grpcdb/server.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	"google.golang.org/grpc/credentials"
+
+	"github.com/tendermint/tmlibs/db"
+	protodb "github.com/tendermint/tmlibs/proto"
+)
+
+// ListenAndServe is a blocking function that sets up a gRPC based
+// server at the address supplied, with the gRPC options passed in.
+// Normally in usage, invoke it in a goroutine like you would for http.ListenAndServe.
+func ListenAndServe(addr string, cert string, key string, opts ...grpc.ServerOption) error {
```
Comment (melekes) [inline grpcdb/server.go]: same here `addr, cert, key string`

File: grpcdb/server.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+		return err
+	}
+	srv, err := NewServer(cert, key, opts...)
+	if err != nil {
+		return err
+	}
+	return srv.Serve(ln)
+}
+
+func NewServer(cert string, key string, opts ...grpc.ServerOption) (*grpc.Server, error) {
```
Comment (melekes) [inline grpcdb/server.go]: same

File: grpcdb/server.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+func (s *server) Delete(ctx context.Context, in *protodb.Entity) (*protodb.Nothing, error) {
+	s.db.Delete(in.Key)
+	return nothing, nil
+}
+
+var nothing = new(protodb.Nothing)
+
+func (s *server) DeleteSync(ctx context.Context, in *protodb.Entity) (*protodb.Nothing, error) {
```
Comment (melekes) [inline grpcdb/server.go]: why return `nothing`?

File: grpcdb/client.go (hunk 4 above)
Comment (cwgoes) [reply grpcdb/client.go]: Addressed.

File: grpcdb/server.go (hunk 6 above)
Comment (cwgoes) [reply grpcdb/server.go]: Addressed.

File: grpcdb/server.go (hunk 7 above)
Comment (cwgoes) [reply grpcdb/server.go]: Addressed.

File: grpcdb/server.go (hunk 8 above)
Comment (cwgoes) [reply grpcdb/server.go]: gRPC needs a value, maybe?

File: grpcdb/doc.go (hunk 5 above)
Comment (cwgoes) [reply grpcdb/doc.go]: Fixed.

**Comment by cwgoes** [comment]
Batch operations implemented, addressed comments. We might want more tests still.

File: remotedb/remotedb.go [hunk 9]
```diff
@@ -218,5 +220,43 @@ func (itr *iterator) Value() []byte {
 }
 
 func (itr *iterator) Close() {
-	// TODO: Shut down the iterator
+	err := itr.dic.CloseSend()
```
Comment (melekes) [inline remotedb/remotedb.go]: weird method name... does `CloseSend` mean close for sending?

Comment (cwgoes) [reply remotedb/remotedb.go]: Close the send-side, I think. Modeled after the example at the bottom of https://grpc.io/docs/tutorials/basic/go.html.

**Comment by odeke-em** [comment]
Great, thank you @cwgoes for shipping this and @melekes for the review!

File: grpcdb/server.go (hunk 8 above)
Comment (odeke-em) [reply grpcdb/server.go]: Yes, the protobuf compiler can't generate an rpc definition that without a return value i.e.
```go
service Foo {
   rpc Bar() 
}
```  
but it'll accept
```go
service Foo {
   rpc Foo(Nothing) returns (Nothing)
}
```

**Pull request merged**